### PR TITLE
Replicaset awareness

### PR DESCRIPTION
**Pull request by KnVerey**
@Shopify/cloudplatform 

This PR makes `KubernetesResource::Deployment` aware of replica sets. This should fix a number of outstanding issues:
- Pod warnings get shown for pods that are being shut down, in the case where the last deploy was bad and the current one is actually succeeding (very confusing output)
- The way we try to identify a representative pod for the purpose of outputting related events and logs on failed deploy is sketchy (https://github.com/Shopify/kubernetes-deploy/pull/98#discussion_r119244951) and could theoretically still select a pod from the wrong RS.
- Rolling updates currently cannot fail, and will never time out on the basis of their pod states ([notes in code](https://github.com/Shopify/kubernetes-deploy/blob/master/lib/kubernetes-deploy/kubernetes_resource/deployment.rb#L62-L67))

Fixes https://github.com/Shopify/kubernetes-deploy/issues/36
Fixes https://github.com/Shopify/kubernetes-deploy/issues/42 (with small amount of extra code + test)

## Problem

We currently make the link between a deployment and its pods by looking for a "name" label. This is super fragile, since we're not using the full selector set, plus the presence of that label is a Shopify convention, not a k8s requirement. Furthermore, even using the full selector set does not guarantee you're looking at the right pods if you've gotten yourself into a really bad state (selector overlap). Ideally we'd connect Deployments to the pods we care about in a robust way; this is much easier to do as of Kubernetes 1.6 (see "Additional context" below).

## Solution

This PR uses the full set of label selectors to grab all possibly relevant RS. It then uses the new `ownerReferences` information to select the RS it owns, and the `deployment.kubernetes.io/revision` to select the current RS. That RS in turn uses label selectors and `ownerReferences` to list its pods.

## Additional context

- kubectl itself identifies the new replicaset by calculating and comparing pod template hashes: https://github.com/kubernetes/kubernetes/blob/a9f5065833d782ea93a6f462674fc7d3fb3affe7/pkg/controller/deployment/util/deployment_util.go#L611-L622. I don' know whether they plan to switch to the new attributes I'm using here, but either way, that approach is not feasible for us.
- `ownerReferences` were added to Deployment in 1.6. The [PR](https://github.com/kubernetes/kubernetes/pull/35676) that introduced them mentions they will be used for RS garbage collection. However, the [issue it resolved](https://github.com/kubernetes/kubernetes/issues/33845) mentions use cases like ours here.
- The `deployment.kubernetes.io/revision` annotation is much older (circa 1.2). According to the [PR](https://github.com/kubernetes/kubernetes/pull/19581) introducing it, the rollout history and rollback features leverage it.

### CODE REVIEW

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      @status = @rollout_data.map { |st, num| "#{num} #{st.chop.pluralize(num)}" }.join(", ")
+    end
+
+    def deploy_succeeded?
+      @rollout_data["replicas"].to_i == @rollout_data["availableReplicas"].to_i &&
+      @rollout_data["replicas"].to_i == @rollout_data["readyReplicas"].to_i
+    end
+
+    def deploy_failed?
+      @pods.present? && @pods.all?(&:deploy_failed?)
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: This (and `deploy_timed_out?`) still will basically never fail on account of the pods, since the Pod timeout is higher and the failure condition is super narrow (`@phase == "Failed"`). However, my next task is to make pod failures smarter / more aggressive, and at that point this should make a huge difference in how often our users hit the Deployment timeout.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 2]
```diff
@@ -4,68 +4,54 @@ class Deployment < KubernetesResource
...
+      @latest_rs.fetch_events
     end
 
-    def fetch_events
-      own_events = super
-      return own_events unless @representative_pod
-      own_events.merge(@representative_pod.fetch_events)
+    def fetch_logs
+      return {} unless @latest_rs.present?
+      @latest_rs.fetch_logs
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: `kubectl get logs deployment/web -c nginx` (or whatever) is a valid command. However, looking at the source, I  think it might favour pods from the old RS under some circumstances, and certainly won't exclude them. https://github.com/kubernetes/kubernetes/blob/dbd1503b6546b0b6048787bad0e63515f3281506/pkg/controller/controller_utils.go#L669-L698

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      unmanaged? ? @found : true
+    end
+
+    def fetch_events
+      own_events = super
+      return own_events unless @pods.present?
+      own_events.merge(@pods.first.fetch_events)
+    end
+
+    def fetch_logs
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: See [the source](https://github.com/kubernetes/kubernetes/blob/dbd1503b6546b0b6048787bad0e63515f3281506/pkg/controller/controller_utils.go#L669-L698) for how kubectl selects which pod to dump logs from. The gist is that it is trying to select a pod that is more likely to actually have logs.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 4]
```diff
@@ -4,68 +4,54 @@ class Deployment < KubernetesResource
...
+          .slice("replicas", "updatedReplicas", "availableReplicas", "unavailableReplicas"))
+        @status = @rollout_data.map { |state_replicas, num| "#{num} #{state_replicas.chop.pluralize(num)}" }.join(", ")
       end
     end
 
-    def fetch_logs
-      @representative_pod ? @representative_pod.fetch_logs : {}
+    def fetch_events
+      return {} unless @latest_rs.present?
+      @latest_rs.fetch_events
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: We could include deployment events in here too, but I don't think they'd generally be useful (typically, they're about "Scaled replicaSet foo"), and there can be a ton of them since `fetch_events` doesn't filter by time right now (though I'm considering adding that).

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      return {} unless container_names.present?
+      container_names.each_with_object({}) do |container_name, container_logs|
+        out, _err, _st = kubectl.run(
+          "logs",
+          id,
+          "--container=#{container_name}",
+          "--since-time=#{@deploy_started.to_datetime.rfc3339}",
+          "--tail=#{LOG_LINE_COUNT}"
+        )
+        container_logs["#{id}/#{container_name}"] = out
```
Comment (mkobetic) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: Given that we're displaying logs for all the containers in the pod, might make sense to display shorter ones first to avoid having a chatty container drown out a less chatty and possibly more relevant one. Visually separating individual container logs would probably be helpful as well.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 6]
```diff
@@ -4,68 +4,54 @@ class Deployment < KubernetesResource
     TIMEOUT = 5.minutes
 
     def sync
-      json_data, _err, st = kubectl.run("get", type, @name, "--output=json")
+      raw_json, _err, st = kubectl.run("get", type, @name, "--output=json")
       @found = st.success?
-      @rollout_data = {}
+      @rollout_data = { "replicas" => 0 }
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Do we expect many calls to `sycn` on the same instance?  or should this be set in the constructor?

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    def initialize(name:, namespace:, context:, file:, parent: nil, logger:)
+      @name = name
+      @namespace = namespace
+      @context = context
+      @file = file
+      @parent = parent
+      @logger = logger
+      @pods = []
+      @rollout_data = { "replicas" => 0 }
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: Oh interesting, this one does do the intialization here, there must be a reason I'm not seeing.

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        rs_data = JSON.parse(raw_json)
+        interpret_json_data(rs_data)
+      else # reset
+        @rollout_data = { "replicas" => 0 }
+        @status = nil
+        @pods = []
+      end
+    end
+
+    def interpret_json_data(rs_data)
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: Does this method need to be public?

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+      pods_data = get_pods(rs_data)
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: I might be argued out of this one, but same as in the RS class, this should probably just return the objects.

File: lib/kubernetes-deploy/kubernetes_resource.rb [hunk 10]
```diff
@@ -192,5 +193,10 @@ def pretty_status
     def kubectl
       @kubectl ||= Kubectl.new(namespace: @namespace, context: @context, logger: @logger, log_failure_by_default: false)
     end
+
+    def template
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource.rb]: Why is this needed, and why is it okay to return `nil` if the file is not present?

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 11]
```diff
@@ -4,68 +4,54 @@ class Deployment < KubernetesResource
...
+        latest_rs_data = get_latest_rs(deployment_data)
+        @latest_rs = ReplicaSet.new(
+          name: latest_rs_data["metadata"]["name"],
+          namespace: namespace,
+          context: context,
+          file: nil,
+          parent: "#{@name.capitalize} deployment",
+          logger: @logger
+        )
+        @latest_rs.interpret_json_data(latest_rs_data)
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: I even think this should be an optional(?) param to the constructor:

`ReplicaSet.new(blah blah... data: json)`

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb (hunk 8 above)
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: I get it now, because the data is set into the objects after intialization, as I mentioned in the other comment, sounds like it could be an optional param to the constructor.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 12]
```diff
@@ -4,68 +4,54 @@ class Deployment < KubernetesResource
...
+      raw_json, _err, st = kubectl.run("get", type, @name, "--output=json")
       @found = st.success?
-      @rollout_data = {}
+      @rollout_data = { "replicas" => 0 }
       @status = nil
-      @representative_pod = nil
-      @pods = []
+      @latest_rs = nil
 
       if @found
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: I know is not a change here, but why is there no defined behaviour for if not found?

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 11 above)
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: That ^ or `@latest_rs = ReplicaSet.from_json(...)`

File: test/fixtures/hello-cloud/bare_replica_set.yml [hunk 13]
```diff
@@ -0,0 +1,27 @@
+---
+apiVersion: extensions/v1beta1
+kind: ReplicaSet
```
Comment (kirs) [inline test/fixtures/hello-cloud/bare_replica_set.yml]: Is there a use case when we would have a manifest for ReplicaSet? I thought it's created automatically by Deployment

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 14]
```diff
@@ -4,68 +4,54 @@ class Deployment < KubernetesResource
...
+        @rollout_data.merge!(deployment_data["status"]
+          .slice("replicas", "updatedReplicas", "availableReplicas", "unavailableReplicas"))
+        @status = @rollout_data.map { |state_replicas, num| "#{num} #{state_replicas.chop.pluralize(num)}" }.join(", ")
       end
     end
 
-    def fetch_logs
-      @representative_pod ? @representative_pod.fetch_logs : {}
+    def fetch_events
+      return {} unless @latest_rs.present?
```
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: I suppose that the deployment would always have a replica set link. Should we raise if it doesn't (catching the possible inconsistent state)?

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 2 above)
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: AFAIK there's no `kubectl get logs`, you probably meant `kubectl logs`. Though I agree that `kubectl logs deployment/web -c web` can be ambiguous.

How did it work before for a deployment with 100 replicas (== pods)? Would `representative_pod` be one of a hundred?

Do we fetch logs for the developers to tell them what went wrong?

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 15]
```diff
@@ -74,10 +60,18 @@ def exists?
...
-      return false unless @deploy_started
-      # Shitty, brittle workaround to identify a pod from the new ReplicaSet before implementing ReplicaSet awareness
-      Time.parse(pod_json["metadata"]["creationTimestamp"]) >= (@deploy_started - 30.seconds)
+    def get_latest_rs(deployment_data)
+      label_string = deployment_data["spec"]["selector"]["matchLabels"].map { |k, v| "#{k}=#{v}" }.join(",")
+      raw_json, _err, st = kubectl.run("get", "replicasets", "--output=json", "--selector=#{label_string}")
+      return unless st.success?
+
+      all_rs_data = JSON.parse(raw_json)["items"]
+      current_revision = deployment_data["metadata"]["annotations"]["deployment.kubernetes.io/revision"]
```
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: It may be handy to use `deployment_data.dig` here (up to you)

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 16]
```diff
@@ -4,68 +4,54 @@ class Deployment < KubernetesResource
...
-      @representative_pod = nil
-      @pods = []
+      @latest_rs = nil
 
       if @found
-        @rollout_data = JSON.parse(json_data)["status"]
-          .slice("updatedReplicas", "replicas", "availableReplicas", "unavailableReplicas")
-        @status = @rollout_data.map { |replica_state, num| "#{num} #{replica_state}" }.join(", ")
+        deployment_data = JSON.parse(raw_json)
+        latest_rs_data = get_latest_rs(deployment_data)
```
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Or `find_latest_rs` since this method actually gets multiple RS and finds the right one.

File: lib/kubernetes-deploy/kubernetes_resource/pod.rb [hunk 17]
```diff
@@ -14,13 +14,13 @@ def initialize(name:, namespace:, context:, file:, parent: nil, logger:)
...
-      out, _err, st = kubectl.run("get", type, @name, "-a", "--output=json")
+      raw_json, _err, st = kubectl.run("get", type, @name, "-a", "--output=json")
       if @found = st.success?
-        pod_data = JSON.parse(out)
+        pod_data = JSON.parse(raw_json)
         interpret_json_data(pod_data)
       else # reset
         @status = @phase = nil
-        @ready = false
+        @ready = @found = false
```
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/pod.rb]: Why do we need the `@found` flag?

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb (hunk 3 above)
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: ❤️ for checking it and for the reference to k8s source that was not easy to find

File: test/integration/kubernetes_deploy_test.rb [hunk 18]
```diff
@@ -381,8 +418,27 @@ def test_output_when_switching_labels_incorrectly
...
-    assert_logs_match("one of your templates is invalid")
-    assert_logs_match(/The Deployment "web" is invalid.*`selector` does not match template `labels`/)
+
+    assert_logs_match_all([
+      "one of your templates is invalid",
+      /The Deployment "web" is invalid.*`selector` does not match template `labels`/
+    ])
+  end
+
+  def test_can_deploy_deployment_with_zero_replicas
```
Comment (kirs) [inline test/integration/kubernetes_deploy_test.rb]: I was going to mention that it's extremely important to make sure that it works with multiple long-running revisions / pods that take more time to shutdown / custom terminationGracePeriod, but it's already tested by `test_long_running_deployment` and if the test passes I don't have much to add.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 2 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Yes, I meant `kubectl logs`. And yes, we're only fetching logs as debug information for these resource types when they fail (unlike unmanaged pods, for which we always dump logs). `representative_pod` was a hack to be able to introduce the debug output without requiring this PR to be ready in addition to the extensive changes that did go in at the time. It attempted to select a pod from the new replicaSet by looking at pod creationTimestamps.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 4 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: My other PR adds filtering by time, so we can actually keep these events after all.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 6 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Yes, `sync` will be called many, many times on the same instance--about once every three seconds from the beginning of the deploy until the time the resource succeeds/fails. The exception to this is resources that were created to model subresources of the thing that was actually in the template set, e.g. managed replica sets or pods. We'll get new instances of those on every sync.

File: lib/kubernetes-deploy/kubernetes_resource.rb (hunk 10 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource.rb]: I put it here because multiple subclasses in this PR use it:
- `ReplicaSet` uses it to grab the container names in the case where it was not derived from a deployment (managed ReplicaSets will not have a template file)
- `Service` uses it to find the related deployment so it can be smarter about how many endpoints to expect (notably so that it can expect zero when appropriate)

Your comment make me think that perhaps I should move the fallback right into this method: if we have a file, load it from file; if we don't, fetch it from the API.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 12 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: The behaviour in that case is a reset, which is accomplished by the code above. Basically, we always reset, and then if we find new data, we repopulate using it.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 16 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: 👍

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 11 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: I remember trying the `ReplicaSet.from_json` model at one point in this work and running into difficulties... but I don't remember what they were, so I'll try again. :P

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 14 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: No, I don't think we should raise. For one, the purpose of `fetch_events` is collection of debug information, so raising during that would throw a wrench in our attempt to be helpful. For another, since the RS creation happens asynchronously, it is theoretically possible for the RS not to exist (yet) at a given point.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 15 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: That's a Ruby 2.3 feature, so I can't (unless we decide to break with the version supported by shipit-engine, or that gem bumped its requirement and I didn't notice)

File: lib/kubernetes-deploy/kubernetes_resource/pod.rb (hunk 17 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/pod.rb]: It's used by `exists?`... but I'm not sure why I moved it here if that's what you mean. It should already be set, so I'll change it back.

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb (hunk 7 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: The reason has to do with the possible origins of these resource types:
- Deployments _always_ come from a file in your deploy directory and get their observed state via `sync`.
- ReplicaSets _usually_ show up via deployments, but it is possible to deploy them separately, so I've supported that here (and added a test).
- Pods are very commonly created both ways: unmanaged from a file in your deploy directory, or via a deployment/replicaSet.

For deployments, we have it easy: we see the file, create the instance, and then call sync on it repeatedly to reset and update the observed state. For the other two, there is the additional possibility that they are being created when their parent syncs, and that the parent already has the data on their observed state. In that case, we create the instance and then call `interpret_json_data` with that data; `sync` is never called. Effectively, `interpret_json_data` is the common core between the two flows. I'll investigate making it private / clarifying these two flows.

File: test/fixtures/hello-cloud/bare_replica_set.yml (hunk 13 above)
Comment (KnVerey) [inline test/fixtures/hello-cloud/bare_replica_set.yml]: I haven't encountered a use case, but it is definitely possible to create them separately.

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb (hunk 5 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: How about this?

image

- each gets its own section, with a heading instead of tags on every line
- sorted by line count
- identifier format `container 'foo'` instead of the full resource path (`Deployment/web/foo`)

File: lib/kubernetes-deploy/kubernetes_resource.rb [hunk 19]
```diff
@@ -139,10 +140,11 @@ def debug_message
...
         else
-          helpful_info << "  - Logs (last #{LOG_LINE_COUNT} lines shown):"
-          container_logs.each do |identifier, logs|
-            logs.split("\n").each do |line|
-              helpful_info << "      [#{identifier}]\t#{line}"
+          sorted_logs = container_logs.sort_by { |_, log_lines| log_lines.length }
+          sorted_logs.each do |identifier, log_lines|
+            helpful_info << "  - Logs from container '#{identifier}' (last #{LOG_LINE_COUNT} lines shown):"
+            log_lines.each do |line|
+              helpful_info << "      #{line}"
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource.rb]: Now looks like this:
image
[...event noise addressed by other PR...]
image

File: lib/kubernetes-deploy/kubernetes_resource/pod.rb [hunk 20]
```diff
@@ -4,50 +4,38 @@ class Pod < KubernetesResource
...
+        if @deploy_started
+          log_suspicious_states(pod_data["status"].fetch("containerStatuses", []))
+        end
+      else # reset
+        @found = false
+        @containers = []
+        @phase = @status = nil
+        @ready = false
       end
+      display_logs if unmanaged? && deploy_succeeded?
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/pod.rb]: Example of this output:

image

**Comment by KnVerey** [comment]
I made some architectural changes to this PR based on your questions and feedback. Now, whether the resource was created from a deploy template or during a parent resource's `sync`, its observed state is recorded through its own `sync` method.`interpret_json_data` is gone. This way, we have a more unified code path for the resources that can be created both ways (i.e. Pods, ReplicaSets). I'm hoping this will be a lot clearer--lmk what you think!

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 21]
```diff
@@ -74,10 +51,27 @@ def exists?
...
+      end
+
+      ReplicaSet.new(
+        name: latest_rs_data["metadata"]["name"],
+        namespace: namespace,
+        context: context,
+        parent: "#{@name.capitalize} deployment",
+        logger: @logger,
+        deploy_started: @deploy_started
+      ).tap { |rs| rs.sync(latest_rs_data) }
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: I can also make a method like this if you still think it is better:
```ruby
    def self.from_json(**args)
      json_data = args.delete(:json_data)
      inst = new(**args)
      inst.sync(json_data)
      inst
    end
```
I'm not sure it adds much value, since most of the initializer params aren't from the json, but I'm not against it.

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb (hunk 5 above)
Comment (mkobetic) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: ❤️ 👍

File: lib/kubernetes-deploy/kubernetes_resource/service.rb [hunk 22]
```diff
@@ -16,7 +16,11 @@ def sync
     end
 
     def deploy_succeeded?
-      @num_endpoints > 0
+      if related_deployment_replicas
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/service.rb]: The purpose of this change is to be able to handle zero-replica deployments (that have been temporarily scaled down for some reason). I'm second-guessing the what I've done here though--I think it might be too aggressive. If the deployment has rolled out and the service has _some_ endpoints, it's not really valuable to hold up the deploy, is it?

WDTY think about this logic instead:

```ruby
if exposes_zero_replica_deployment?
 @num_endpoints == 0
else
  @num_endpoints > 0
end
```

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 21 above)
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: I think is fine as is, what I don't like is `tap` I find that method to be kinda not very useful sintactic sugar; WDYJ:

```
rs = RS.new(..)
rs.sync
```

File: lib/kubernetes-deploy/kubernetes_resource/pod.rb [hunk 23]
```diff
@@ -4,50 +4,38 @@ class Pod < KubernetesResource
...
     SUSPICIOUS_CONTAINER_STATES = %w(ImagePullBackOff RunContainerError ErrImagePull).freeze
 
-    def initialize(name:, namespace:, context:, file:, parent: nil, logger:)
+    def initialize(name:, namespace:, context:, file: nil, parent: nil, logger:, deploy_started: nil)
       @name = name
       @namespace = namespace
       @context = context
       @file = file
       @parent = parent
+      @deploy_started = deploy_started
```
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/pod.rb]: This is an interesting semantic change here; why would a `KubernetesResource` be aware of when a deploy started?

File: lib/kubernetes-deploy/kubernetes_resource/service.rb (hunk 22 above)
Comment (camilo) [inline lib/kubernetes-deploy/kubernetes_resource/service.rb]: I like the explicitness of `exposes_zero_replica_deployment?` over the current more implicit implementation

File: lib/kubernetes-deploy/kubernetes_resource.rb (hunk 10 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource.rb]: Memoizing _should_ be safe... From file, the result obviously won't change. Although resources do change cluster-side, the ones that don't have a file are child resources for which that shouldn't be true. For example, if the initial status check (before the deploy makes any changes) on a deployment resource memoized a template from the cluster, that'd be bad, since that template would change when the deploy proceeds. But the child resource objects are replaced by their parents every time the parent syncs.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 24]
```diff
@@ -4,68 +4,45 @@ class Deployment < KubernetesResource
...
-
-            if !@representative_pod && pod_probably_new?(pod_json)
-              @representative_pod = pod
-            end
-            @pods << pod
-          end
-        end
+        deployment_data = JSON.parse(raw_json)
+        @latest_rs = find_latest_rs(deployment_data)
+        @rollout_data = { "replicas" => 0 }.merge!(deployment_data["status"]
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: no good reason. I refactored this section a bunch of times, so maybe there was some intermediate version that didn't use the return value. I'll fix it.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 21 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Because we need to return the rs, so it'd be _three_ lines instead of one 😉 
Though our style guide doesn't mention it, I've heard other Shopifolk hating on `tap` in the past. I can change it.

File: lib/kubernetes-deploy/kubernetes_resource/pod.rb (hunk 23 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/pod.rb]: It's not actually new. I added it as a parameter to save the parent resource from having to call the setter immediately in these cases. I'll change it back if you think this introduces confusion. See [here in the diff](https://github.com/Shopify/kubernetes-deploy/pull/107/files#diff-4709b5257970978173e422904663525fL32) for the setter being used instead.

As for why it is needed, it is primarily so that they know when to timeout, although we also use it to filter logs and events. It's not the global deploy start time; it is the time that specific resource was `apply`'d. For resources in your yamls, the instance is created when we first read in the templates, and this attribute is it is set later,  [here](https://github.com/Shopify/kubernetes-deploy/blob/b61fe310824a17ce82f57964bc44d641ffc6c169/lib/kubernetes-deploy/runner.rb#L376).

**Comment by KnVerey** [comment]
@camilo @kirs This has been rebased on the refactor I extracted out of it based on your feedback. What's left should look pretty familiar, but could you please take a final look? 😄

**Comment by KnVerey** [comment]
@ibawt can you please review this one?

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+    end
+
+    def desired_replicas
+      @definition["spec"]["replicas"].to_i
+    end
+
+    def fetch_events
+      own_events = super
+      return own_events unless @pods.present?
+      own_events.merge(@pods.first.fetch_events)
```
Comment (ibawt) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: pods.first: does that mean we'll lose some data or is it the same for the entire thing?

Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: My thinking was that the events will be very similar across the entire set of pods. It is possible some pods will be more interesting than others (e.g. the case you saw where a pod landed on a bad node), but at this stage I don't think we have a nice way of figuring out which one. After my next PR, we could deliberately try to find a failed pod (if any) and use its events... 🤔  Did you have any ideas around this?

File: test/integration/kubernetes_deploy_test.rb [hunk 26]
```diff
@@ -384,8 +422,27 @@ def test_output_when_switching_labels_incorrectly
...
+    success = deploy_fixtures("hello-cloud", subset: ["configmap-data.yml", "web.yml.erb"]) do |fixtures|
+      web = fixtures["web.yml.erb"]["Deployment"].first
+      web["spec"]["replicas"] = 0
```
Comment (kirs) [inline test/integration/kubernetes_deploy_test.rb]: A use case of zero replicas that I remember is:

1. Create a deployment with unspecified `replicas` field
2. The replicas number is scaled dynamically (can sometimes be zero)
3. At the moment of the deploy it's zero

I see the test works with YAML that explicitly sets `replicas` to zero, but would it also support the case of dynamic `replicas` value set to zero?

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 27]
```diff
@@ -4,78 +4,84 @@ class Deployment < KubernetesResource
...
+      label_string = deployment_data["spec"]["selector"]["matchLabels"].map { |k, v| "#{k}=#{v}" }.join(",")
+      raw_json, _err, st = kubectl.run("get", "replicasets", "--output=json", "--selector=#{label_string}")
+      return unless st.success?
+
+      all_rs_data = JSON.parse(raw_json)["items"]
+      current_revision = deployment_data["metadata"]["annotations"]["deployment.kubernetes.io/revision"]
+
+      latest_rs_data = all_rs_data.find do |rs|
+        rs["metadata"]["ownerReferences"].any? { |ref| ref["uid"] == deployment_data["metadata"]["uid"] } &&
+        rs["metadata"]["annotations"]["deployment.kubernetes.io/revision"] == current_revision
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: I'm using `ownerReferences` to be completely certain that the RS belongs to the deployment, and `revision` to select the new one, as opposed to the `n` other RS that could exists for this deployment. This seemed like the best available option based one the k8s source and issues I read, but I could be missing something. Are you aware of a better way?

File: test/integration/kubernetes_deploy_test.rb (hunk 26 above)
Comment (KnVerey) [inline test/integration/kubernetes_deploy_test.rb]: Yes, that should work fine. [The implementation](https://github.com/Shopify/kubernetes-deploy/pull/107/files#diff-bfab9900d04f0a61190aa04f896aed1fR50) uses selectors to grab a live copy of the related deployment's template, so it should work even if the deployment wasn't managed by `kubernetes-deploy`.

Comment (kirs) [inline test/integration/kubernetes_deploy_test.rb]: :+1: great work!

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 27 above)
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Ah, I somehow thought it's the k8s revision

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb (hunk 25 above)
Comment (ibawt) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: no ideas just wanted to make sure I knew the reason.
